src/server/game/AI/SmartScripts/SmartScript.cpp

```diff
@@ -1017,16 +1017,24 @@ void SmartScript::ProcessAction(SmartScriptHolder& e, Unit* unit, uint32 var0, u
 
             for (ObjectList::const_iterator itr = targets->begin(); itr != targets->end(); ++itr)
             {
-                if (!IsCreature(*itr))
+                if (!IsCreature(*itr) || !IsGameObject(*itr))
                     continue;
 
-                if ((*itr)->ToUnit()->IsAlive() && IsSmart((*itr)->ToCreature()))
+                if (Creature* target = (*itr)->ToCreature())
                 {
-                    ENSURE_AI(SmartAI, (*itr)->ToCreature()->AI())->SetDespawnTime(e.action.forceDespawn.delay + 1); // Next tick
-                    ENSURE_AI(SmartAI, (*itr)->ToCreature()->AI())->StartDespawn();
+                    if (target->IsAlive() && IsSmart(target))
+                    {
+                        ENSURE_AI(SmartAI, target->AI())->SetDespawnTime(e.action.forceDespawn.delay + 1); // Next tick
+                        ENSURE_AI(SmartAI, target->AI())->StartDespawn();
+                    }
+                    else
+                        target->DespawnOrUnsummon(e.action.forceDespawn.delay);
+                }
+                else if (GameObject* goTarget = (*itr)->ToGameObject())
+                {
+                    if (IsSmartGO(goTarget))
+                        goTarget->SetRespawnTime(e.action.forceDespawn.delay + 1);
```

Comment:
Line can be removed if you want to allow removing GOs without SmartScripts, I'm not coding DB stuff so I don't know if this is a good idea or it is used in any way.